src/palimpzest/execution/execution_engine.py

```diff
@@ -0,0 +1,333 @@
+        self.available_models = available_models
+        if not available_models:
+            self.available_models = getModels()
+        if self.verbose:
+            print("Available models: ", self.available_models)
+        self.allow_bonded_query = allow_bonded_query
+        self.allow_conventional_query = allow_conventional_query
+        self.allow_model_selection = allow_model_selection
+        self.allow_code_synth = allow_code_synth
+        self.allow_token_reduction = allow_token_reduction
+        self.optimization_strategy = optimization_strategy
+        self.pareto_strategy_min_trials = pareto_strategy_min_trials
+        self.max_workers = max_workers
+        self.num_workers_per_thread = num_workers_per_thread
+        self.inter_plan_parallelism = inter_plan_parallelism
+        
+        self.datadir = DataDirectory()
+
+
+    def execution_id(self) -> str:
+        """
+        Hash of the class parameters.
+        """
+        id_str = ""
+        for attr, value in self.__dict__.items():
+            if not attr.startswith("_"):
+                id_str += f"{attr}={value},"
+
+        return hashlib.sha256(id_str.encode("utf-8")).hexdigest()[:MAX_ID_CHARS]
+
+
+    def clear_cache(self):
+        """
+        Clear cached LLM responses and codegen samples.
+        """
+        dspy_cache_dir = os.path.join(os.path.expanduser("~"), "cachedir_joblib/joblib/dsp/")
+        if os.path.exists(dspy_cache_dir):
+            shutil.rmtree(dspy_cache_dir)
+        cache = self.datadir.getCacheService()
+        cache.rmCache()
+
+
+    def set_source_dataset_id(self, dataset: Set) -> str:
+        """
+        Sets the dataset_id of the DataSource for the given dataset.
+        """
+        # iterate until we reach DataSource
+        while isinstance(dataset, Set):
+            dataset = dataset._source
+
+        # throw an exception if datasource is not registered with PZ
+        _ = self.datadir.getRegisteredDataset(dataset.dataset_id)
+
+        # set the source dataset id
+        self.source_dataset_id = dataset.dataset_id
+
+
+    def get_parallel_max_workers(self):
+        # for now, return the number of system CPUs;
+        # in the future, we may want to consider the models the user has access to
+        # and whether or not they will encounter rate-limits. If they will, we should
+        # set the max workers in a manner that is designed to avoid hitting them.
+        # Doing this "right" may require considering their logical, physical plan,
+        # and tier status with LLM providers. It may also be worth dynamically
+        # changing the max_workers in response to 429 errors.
+        return multiprocessing.cpu_count()
+
+
+    def get_max_quality_plan_id(self, plans: List[PhysicalPlan]) -> str:
+        """
+        Return the plan_id for the plan with the highest quality in the list of plans.
+        """
+        max_quality_plan_id, max_quality = None, -1
+        for plan in plans:
```

Comment:
Here's an idea: can we store plans sorted by their quality? (This should also somewow implicitly sort by cost)
Therefore if we have a lot of plans we dont need a full for loop. But if we are not hitting significant bottlenecks maybe let's leave it as it is.

Reply:
We can definitely do that -- right now the # of plans is very small so I will leave as-is, but definitely something we can do in a subsequent PR